frame/clearing-house/src/lib.rs

```diff
@@ -812,30 +798,84 @@ pub mod pallet {
 				let pnl = exit_value.try_sub(&entry_value)?;
 				// Realize PnL
 				// TODO(0xangelo): properly handle bad debt incurred by large negative PnL
-				AccountsMargin::<T>::insert(
-					account_id,
-					Self::update_margin_with_pnl(&margin, &pnl)?,
-				);
+				margin = Self::update_margin_with_pnl(&margin, &pnl)?;
 			}
 
+			// Charge fees
+			let fee = Self::fee_for_trade(&market, &quote_abs_amount_decimal)?;
+			margin = margin.checked_sub(&fee).ok_or(ArithmeticError::Underflow)?;
+			market.fee_pool = market.fee_pool.checked_add(&fee).ok_or(ArithmeticError::Overflow)?;
+
+			// Check account risk
 			if is_risk_increasing {
 				ensure!(
 					Self::is_above_imr(&positions, margin)?,
 					Error::<T>::InsufficientCollateral
 				);
 			}
 
+			// Update storage
+			AccountsMargin::<T>::insert(account_id, margin);
 			Positions::<T>::insert(account_id, positions);
+			Self::update_market_after_trade(&mut market, base_swapped, direction)?;
+			Markets::<T>::insert(market_id, market);
 
 			Self::deposit_event(Event::TradeExecuted {
 				market: market_id.clone(),
 				direction,
 				quote: quote_asset_amount,
 				base: base_swapped,
 			});
-
 			Ok(base_swapped)
 		}
+
+		#[transactional]
+		fn update_funding(market_id: &Self::MarketId) -> Result<(), DispatchError> {
+			let mut market = Markets::<T>::get(market_id).ok_or(Error::<T>::MarketIdNotFound)?;
+
+			// Check that enough time has passed since last update
+			let now = T::UnixTime::now().as_secs();
+			ensure!(
+				now - market.funding_rate_ts >= market.funding_frequency,
+				Error::<T>::UpdatingFundingTooEarly
+			);
+
+			// Pay funding
+			// net position sign | funding rate sign | transfer
+			// --------------------------------------------------------------
+			//                -1 |                -1 | Collateral -> IF
+			//                -1 |                 1 | Fee Pool -> Collateral
+			//                 1 |                -1 | Fee Pool -> Collateral
+			//                 1 |                 1 | Collateral -> IF
+			//                 - |                 0 | n/a
+			//                 0 |                 - | n/a
```

Comment:
This is wrong actually. All transfers should be between the Collateral and Fee Pool. Maybe I can change this in a future PR?